### PR TITLE
Fix 500 Error with branch and tag sharing the same name

### LINKED ISSUE #15592: PR Error 500 with branch and tag sharing the same name
- Gitea version (or commit ref): 1.14.1
- Git version: 2.30.2
- Operating system: docker image gitea/gitea:1

- Database:
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite

- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes: https://try.gitea.io/berjou/same-tag-and-branch-name-test/pulls

  - [ ] No
  
- Log gist:
Sorry, tinkering a lot with my setup right now, the log are no longer present on my machine, and since the issue is easily reproductible, it should not be a problem. However, I'm happy to provide you specific logs if needed.

## Description
- `git checkout -b same-tag-and-branch-name-test`
- Create a commit and `git push origin same-tag-and-branch-name-test`
- `git tag same-tag-and-branch-name-test`
- `git push origin tag same-tag-and-branch-name-test`
- Create a pull request to merge `same-tag-and-branch-name-test` into `master`
- Now `https://gitea-instance/user/repo/pulls` return an error 500
- Note that we can see the pull request by accessing `https://gitea-instance/user/repo/1`

Notes:
After deleting the branch and tag, pushing my commit to a new brarnch with a different name, and opening a new pull request, the error 500 on `https://gitea-instance/user/repo/pulls` is gone. However, after merging the merge request, and now that there is no more opened PR, the error 500 came back.

Hope this issue can be fixed, and that someone could help me to fix my broken repository.

### PR DESCRIPTION
**Pull request by Viktor-Yakovchuk**
Fixed 500 error while create Pull request when there are more
than one sources (branch, tag) with the same name
The idea is to take branch instead of tag by default if the branch and tag name are equal

Fix #15592

Signed-off-by: Viktor Yakovchuk <viktor@yakovchuk.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Hmm... Do we ever not want to push to refs/heads? I suspect we should be making this determination before push fails in the first place. 

I'll have a look.

**Comment by Viktor-Yakovchuk** [comment]
> Hmm... Do we ever not want to push to refs/heads? I suspect we should be making this determination before push fails in the first place.
> 
> I'll have a look.

Yes we want to push without "refs/heads" when we create Pull request from tag instead branch. If we always push with refs/heads we will be not able to create pull request from tag.

**Comment by zeripath** [comment]
I would guess that we should be setting the correct ref in the PR instead.

**Comment by Viktor-Yakovchuk** [comment]
As I understand it, you will have to rewrite a lot, especially the routing system, to understand from tag or branch we do Pull request. And as I see GitHub use similar routes and take branch as default when the branch and tag names are equal.

**Comment by zeripath** [comment]
I think the pr.HeadBranch and pr.BaseBranch should have had the refs/heads/... in them and I am fairly certain that at some point I forcibly added this. Somewhere along the line this appears to have been dropped - which is fairly disappointing as this is precisely why I would have added this.

**Comment by Viktor-Yakovchuk** [comment]
Hello, it seems to me that we have 2 options for solving the #15592:

- this patch, perhaps with additional changes recommended by you
- adding refs/heads wherever such operations occur

Еhe second option makes it impossible to create a PR from the tag, but I can see from code and interface that this functionality is planned, although it does not work yet.

I am ready to implement both options in a few days, which one should I choose?

**Comment by Viktor-Yakovchuk** [comment]
@zeripath @lunny guys which way we choose to solve this problem?
1. We delete tags tab from source selector when create pull request (in all over cases we keep all as it is) sea also #15695 Also add "refs/heads" to the code.
2. We take my pull request and fix pull request from tag (now it doesn't work)

image

I am ready to prepare new PR after we choose a way to solve the problem.

**Comment by zeripath** [comment]
OK you have a working fix that fixes the problem for now so I suspect we should merge this.

However, my suggestion is that we should be putting the correct thing in the PR from the start.

So, if you are creating a PR from a branch - then it gets the full ref name: refs/heads/branch.

If you are creating a PR from a tree - it's the full sha

If it's a tag then it's refs/tags/xxx

And so on.

The PR has to be correct and correct forever.

if we use aliases then as you've shown it breaks everything because the name is not a full unambiguous ref.

**Review by zeripath — APPROVED** [review]
This is not the ideal and correct way to fix this but it does at least ameliorate this issue.

We should move to putting the fullref in the PR table so that the situation is always unambiguous.

That would allow us to have PRs opened from remote branches and tags more easily etc.

**Comment by Viktor-Yakovchuk** [comment]
@techknowlogick  please take a look once time permit.

**Review by techknowlogick — APPROVED** [review]

File: services/pull/pull.go [hunk 1]
```diff
@@ -427,6 +431,14 @@ func PushToBaseRepo(pr *models.PullRequest) (err error) {
 			rejectErr := err.(*git.ErrPushRejected)
 			log.Info("Unable to push PR head for %s#%d (%-v:%s) due to rejection:\nStdout: %s\nStderr: %s\nError: %v", pr.BaseRepo.FullName(), pr.Index, pr.BaseRepo, gitRefName, rejectErr.StdOut, rejectErr.StdErr, rejectErr.Err)
 			return err
+		} else if git.IsErrMoreThanOne(err) {
+			log.Info("Retrying to push with refs/heads/%s", pr.HeadBranch)
+			if prefixHeadBranch != "" {
+				log.Info("Can't push with refs/heads/%s", pr.HeadBranch)
```
Comment (6543) [inline services/pull/pull.go]:
```suggestion
				log.Info("Can't push with %s%s", prefixHeadBranch, pr.HeadBranch)
```

File: services/pull/pull.go [hunk 2]
```diff
@@ -427,6 +431,14 @@ func PushToBaseRepo(pr *models.PullRequest) (err error) {
...
+				return err
+			}
+			err = pushToBaseRepoHelper(pr, "refs/heads/")
```
Comment (6543) [inline services/pull/pull.go:440]:
```suggestion
			log.Info("Retrying to push with refs/heads/%s", pr.HeadBranch)
			err = pushToBaseRepoHelper(pr, "refs/heads/")
```

File: services/pull/pull.go [hunk 3]
```diff
@@ -427,6 +431,14 @@ func PushToBaseRepo(pr *models.PullRequest) (err error) {
 			rejectErr := err.(*git.ErrPushRejected)
 			log.Info("Unable to push PR head for %s#%d (%-v:%s) due to rejection:\nStdout: %s\nStderr: %s\nError: %v", pr.BaseRepo.FullName(), pr.Index, pr.BaseRepo, gitRefName, rejectErr.StdOut, rejectErr.StdErr, rejectErr.Err)
 			return err
+		} else if git.IsErrMoreThanOne(err) {
+			log.Info("Retrying to push with refs/heads/%s", pr.HeadBranch)
```
Comment (6543) [inline services/pull/pull.go]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Pull request merged**